File: pages/risk_fault.js
(function () {
  const response = {
    "type": "page",
    "body": [
      {
        "type": "button",
        "label": "新增风险与故障记录",
        "icon": "fa fa-plus",
        "level": "primary",
        "className": "m-b-sm",
        "actionType": "dialog",
        "dialog": {
          "title": "新增风险与故障记录",
          "body": {
            "type": "form",
            "api": {
              "method": "post",
              "url": "/risk_fault/risk_fault",
              "messages": {
                "success": "添加成功",
                "failed": "添加失败"
              }
            },
            "body": [
              {"type": "input-text", "name": "connect_key", "label": "盾构机编号", "required": true},
              {"type": "input-number", "name": "ring", "label": "施工环号", "required": true},
              {"type": "input-text", "name": "sub_code", "label": "施工区间编码", "required": true},
              {"type": "input-text", "name": "sub_name", "label": "施工区间名称", "required": true},
              {"type": "input-text", "name": "risk_start_end_ring", "label": "风险源起止环号"},
              {"type": "input-text", "name": "risk_type", "label": "风险类型"},
              {"type": "input-text", "name": "risk_content", "label": "风险源信息"},
              {"type": "input-text", "name": "risk_level", "label": "风险等级"},
              {"type": "input-text", "name": "risk_desc", "label": "风险描述"},
              {"type": "input-text", "name": "risk_prevention", "label": "预防措施"},
              {"type": "input-text", "name": "risk_treatment", "label": "处理方法"},
              {"type": "input-text", "name": "risk_impact", "label": "对工程的影响"},
              {"type": "input-text", "name": "risk_early_warning", "label": "风险预警指标"},
              {"type": "input-text", "name": "risk_early_warning_value", "label": "风险预警值"},
              {"type": "input-text", "name": "risk_monitoring_frequency", "label": "监测频率"},
              {"type": "input-text", "name": "surface_settlement_monitoring_value", "label": "地表沉降监测值"},
              {"type": "input-text", "name": "structure_deformation_value", "label": "建构筑物变形值"},
              {"type": "input-text", "name": "alarm_ring_section", "label": "警报所在环号"},
              {"type": "input-text", "name": "alarm_type", "label": "警报类型"},
              {"type": "input-text", "name": "fault_occurrence_time", "label": "故障发生时间"},
              {"type": "input-text", "name": "specific_fault_equipment", "label": "故障具体设备"},
              {"type": "input-text", "name": "fault_type", "label": "故障类型"},
              {"type": "input-text", "name": "fault_treatment_measures", "label": "故障处理措施"},
              {"type": "input-text", "name": "sg_fssj", "label": "事故发生时间"},
              {"type": "input-text", "name": "sg_yy", "label": "事故原因"},
              {"type": "input-text", "name": "cl_jg", "label": "处理结果"},
              {"type": "input-text", "name": "wx_yy", "label": "维修原因"},
              {"type": "input-text", "name": "wx_nr", "label": "维修内容"}
            ]
          }
        }
      },
      {
        "type": "crud",
        "syncLocation": false,
        "resizable": true,
        "columnsTogglable": true,
        "className": "table-db m-b-none",
        "tableClassName": "table-cell-nowrap",
        "api": {
          "url": "/risk_fault/risk_fault_list",
          "method": "get",
          "dataType": "json"
        },
        "headerToolbar": [
          {
            "type": "tpl",
            "tpl": "<div class='header-title'><i class='fa fa-cog'></i> 风险与故障记录管理</div>"
          },
          "export-excel",
          {
            "type": "reload",
            "icon": "fa fa-refresh",
            "tooltip": "刷新数据",
            "tooltipPlacement": "top",
            "level": "link"
          }
        ],
        "filter": {
          "title": "查询条件",
          "submitText": "查询",
          "body": [
            {"type": "input-text", "name": "sub_code", "label": "施工区间编码", "placeholder": "请输入施工区间编码进行查询"},
            {"type": "input-text", "name": "sub_name", "label": "施工区间名称", "placeholder": "请输入施工区间名称进行查询"}
          ]
        },
        "columns": [
          {
            "type": "operation",
            "label": "操作",
            "width": 80,
            "fixed": "right",
            "buttons": [
              {
                "type": "button",
                "icon": "fa fa-pencil",
                "tooltip": "修改",
                "level": "link",
                "actionType": "dialog",
                "dialog": {
                  "title": "修改风险与故障记录",
                  "body": {
                    "type": "form",
                    "api": {
                      "method": "post",
                      "url": "/risk_fault/update_risk_fault",
                      "messages": {
                        "success": "更新成功",
                        "failed": "更新失败"
                      }
                    },
                    "body": [
                      {"type": "hidden", "name": "id"},
                      {"type": "input-text", "name": "connect_key", "label": "盾构机编号", "required": true},
                      {"type": "input-number", "name": "ring", "label": "施工环号", "required": true},
                      {"type": "input-text", "name": "sub_code", "label": "施工区间编码", "required": true},
                      {"type": "input-text", "name": "sub_name", "label": "施工区间名称", "required": true},
                      {"type": "input-text", "name": "risk_start_end_ring", "label": "风险源起止环号"},
                      {"type": "input-text", "name": "risk_type", "label": "风险类型"},
                      {"type": "input-text", "name": "risk_content", "label": "风险源信息"},
                      {"type": "input-text", "name": "risk_level", "label": "风险等级"},
                      {"type": "input-text", "name": "risk_desc", "label": "风险描述"},
                      {"type": "input-text", "name": "risk_prevention", "label": "预防措施"},
                      {"type": "input-text", "name": "risk_treatment", "label": "处理方法"},
                      {"type": "input-text", "name": "risk_impact", "label": "对工程的影响"},
                      {"type": "input-text", "name": "risk_early_warning", "label": "风险预警指标"},
                      {"type": "input-text", "name": "risk_early_warning_value", "label": "风险预警值"},
                      {"type": "input-text", "name": "risk_monitoring_frequency", "label": "监测频率"},
                      {"type": "input-text", "name": "surface_settlement_monitoring_value", "label": "地表沉降监测值"},
                      {"type": "input-text", "name": "structure_deformation_value", "label": "建构筑物变形值"},
                      {"type": "input-text", "name": "alarm_ring_section", "label": "警报所在环号"},
                      {"type": "input-text", "name": "alarm_type", "label": "警报类型"},
                      {"type": "input-text", "name": "fault_occurrence_time", "label": "故障发生时间"},
                      {"type": "input-text", "name": "specific_fault_equipment", "label": "故障具体设备"},
                      {"type": "input-text", "name": "fault_type", "label": "故障类型"},
                      {"type": "input-text", "name": "fault_treatment_measures", "label": "故障处理措施"},
                      {"type": "input-text", "name": "sg_fssj", "label": "事故发生时间"},
                      {"type": "input-text", "name": "sg_yy", "label": "事故原因"},
                      {"type": "input-text", "name": "cl_jg", "label": "处理结果"},
                      {"type": "input-text", "name": "wx_yy", "label": "维修原因"},
                      {"type": "input-text", "name": "wx_nr", "label": "维修内容"}
                    ]
                  }
                }
              },
              {
                "type": "button",
                "icon": "fa fa-times text-danger",
                "actionType": "ajax",
                "tooltip": "删除",
                "confirmText": "确定要删除该记录吗？",
                "api": {
                  "method": "delete",
                  "url": "/risk_fault/delete_risk_fault/${id}"
                },
                "messages": {
                  "success": "删除成功",
                  "failed": "删除失败"
                }
              }
            ]
          },
          { "name": "id", "label": "编号","type": "number", "width": 80 },
          { "name": "connect_key", "label": "盾构机编号", "type": "text", "width": 120 },
          { "name": "ring", "label": "施工环号", "type": "number", "width": 100 },
          { "name": "sub_code", "label": "施工区间编码", "type": "text", "width": 120 },
          { "name": "sub_name", "label": "施工区间名称", "type": "text", "width": 150 },
          { "name": "risk_start_end_ring", "label": "风险源起止环号", "type": "text", "width": 120 },
          { "name": "risk_type", "label": "风险类型", "type": "text", "width": 100 },
          { "name": "risk_content", "label": "风险源信息", "type": "text", "width": 150 },
          { "name": "risk_level", "label": "风险等级", "type": "text", "width": 100 },
          { "name": "risk_desc", "label": "风险描述", "type": "text", "width": 150 },
          { "name": "risk_prevention", "label": "预防措施", "type": "text", "width": 150 },
          { "name": "risk_treatment", "label": "处理方法", "type": "text", "width": 150 },
          { "name": "risk_impact", "label": "对工程的影响", "type": "text", "width": 150 },
          { "name": "risk_early_warning", "label": "风险预警指标", "type": "text", "width": 120 },
          { "name": "risk_early_warning_value", "label": "风险预警值", "type": "text", "width": 120 },
          { "name": "risk_monitoring_frequency", "label": "监测频率", "type": "text", "width": 100 },
          { "name": "surface_settlement_monitoring_value", "label": "地表沉降监测值", "type": "text", "width": 150 },
          { "name": "structure_deformation_value", "label": "建构筑物变形值", "type": "text", "width": 150 },
          { "name": "alarm_ring_section", "label": "警报所在环号", "type": "text", "width": 120 },
          { "name": "alarm_type", "label": "警报类型", "type": "text", "width": 100 },
          { "name": "fault_occurrence_time", "label": "故障发生时间", "type": "text", "width": 150 },
          { "name": "specific_fault_equipment", "label": "故障具体设备", "type": "text", "width": 150 },
          { "name": "fault_type", "label": "故障类型", "type": "text", "width": 100 },
          { "name": "fault_treatment_measures", "label": "故障处理措施", "type": "text", "width": 150 },
          { "name": "sg_fssj", "label": "事故发生时间", "type": "text", "width": 150 },
          { "name": "sg_yy", "label": "事故原因", "type": "text", "width": 150 },
          { "name": "cl_jg", "label": "处理结果", "type": "text", "width": 150 },
          { "name": "wx_yy", "label": "维修原因", "type": "text", "width": 150 },
          { "name": "wx_nr", "label": "维修内容", "type": "text", "width": 150 }
        ]
      }
    ]
  };

  window.jsonpCallback && window.jsonpCallback(response);
})();

File: api/ORM/api/risk_fault.py
from fastapi import APIRouter, Query, HTTPException
from typing import Optional
from risk_fault_models import risk_fault_records
from tortoise.expressions import Q
from pydantic import BaseModel

risk_fault_api = APIRouter()

class riskfaultIn(BaseModel):
    connect_key: str
    ring: int
    sub_code: str
    sub_name: str
    risk_start_end_ring: Optional[str] = None
    risk_type: Optional[str] = None
    risk_content: Optional[str] = None
    risk_level: Optional[str] = None
    risk_desc: Optional[str] = None
    risk_prevention: Optional[str] = None
    risk_treatment: Optional[str] = None
    risk_impact: Optional[str] = None
    risk_early_warning: Optional[str] = None
    risk_early_warning_value: Optional[str] = None
    risk_monitoring_frequency: Optional[str] = None
    surface_settlement_monitoring_value: Optional[str] = None
    structure_deformation_value: Optional[str] = None
    alarm_ring_section: Optional[str] = None
    alarm_type: Optional[str] = None
    fault_occurrence_time: Optional[str] = None
    specific_fault_equipment: Optional[str] = None
    fault_type: Optional[str] = None
    fault_treatment_measures: Optional[str] = None
    sg_fssj: Optional[str] = None
    sg_yy: Optional[str] = None
    cl_jg: Optional[str] = None
    wx_yy: Optional[str] = None
    wx_nr: Optional[str] = None
    
@risk_fault_api.post("/risk_fault")
async def add_risk_fault(info_in: riskfaultIn):
    try:
        new_info = await risk_fault_records.create(**info_in.dict(exclude_unset=True))
        return {"status": 0, "msg": "添加成功", "data": {"id": new_info.id}}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"添加失败: {str(e)}")

@risk_fault_api.get("/risk_fault_list")
async def get_risk_fault_list(
    id: Optional[int] = Query(None, description="编号"),
    sub_code: Optional[str] = Query(None, description="施工区间编码"),
    sub_name: Optional[str] = Query(None, description="施工区间名称"),
    page: int = Query(1, ge=1, description="页码"),
    per_page: int = Query(10, ge=1, le=100, description="每页数量")
):
    filters = Q()
    if id is not None:
        filters &= Q(id=id)
    if sub_code:
        filters &= Q(sub_code=sub_code)
    if sub_name:
        filters &= Q(sub_name=sub_name)

    total = await risk_fault_records.filter(filters).count()
    items = await risk_fault_records.filter(filters).offset((page - 1) * per_page).limit(per_page).all()
    
    # 将数据库对象转换为字典列表
    items_dict = [{
        'id': item.id,
        'connect_key': item.connect_key,
        'ring': item.ring,
        'sub_code': item.sub_code,
        'sub_name': item.sub_name,
        'risk_start_end_ring': item.risk_start_end_ring,
        'risk_type': item.risk_type,
        'risk_content': item.risk_content,
        'risk_level': item.risk_level,
        'risk_desc': item.risk_desc,
        'risk_prevention': item.risk_prevention,
        'risk_treatment': item.risk_treatment,
        'risk_impact': item.risk_impact,
        'risk_early_warning': item.risk_early_warning,
        'risk_early_warning_value': item.risk_early_warning_value,
        'risk_monitoring_frequency': item.risk_monitoring_frequency,
        'surface_settlement_monitoring_value': item.surface_settlement_monitoring_value,
        'structure_deformation_value': item.structure_deformation_value,
        'alarm_ring_section': item.alarm_ring_section,
        'alarm_type': item.alarm_type,
        'fault_occurrence_time': item.fault_occurrence_time,
        'specific_fault_equipment': item.specific_fault_equipment,
        'fault_type': item.fault_type,
        'fault_treatment_measures': item.fault_treatment_measures,
        'sg_fssj': item.sg_fssj,
        'sg_yy': item.sg_yy,
        'cl_jg': item.cl_jg,
        'wx_yy': item.wx_yy,
        'wx_nr': item.wx_nr
    } for item in items]

    return {
        "status": 0,
        "msg": "获取成功",
        "data": {
            "items": items_dict,
            "total": total,
            "page": page,
            "perPage": per_page,
            "count": total
        }
    }

class riskfaultUpdateModel(BaseModel):
    id: int
    connect_key: str
    ring: int
    sub_code: str
    sub_name: str
    risk_start_end_ring: Optional[str] = None
    risk_type: Optional[str] = None
    risk_content: Optional[str] = None
    risk_level: Optional[str] = None
    risk_desc: Optional[str] = None
    risk_prevention: Optional[str] = None
    risk_treatment: Optional[str] = None
    risk_impact: Optional[str] = None
    risk_early_warning: Optional[str] = None
    risk_early_warning_value: Optional[str] = None
    risk_monitoring_frequency: Optional[str] = None
    surface_settlement_monitoring_value: Optional[str] = None
    structure_deformation_value: Optional[str] = None
    alarm_ring_section: Optional[str] = None
    alarm_type: Optional[str] = None
    fault_occurrence_time: Optional[str] = None
    specific_fault_equipment: Optional[str] = None
    fault_type: Optional[str] = None
    fault_treatment_measures: Optional[str] = None
    sg_fssj: Optional[str] = None
    sg_yy: Optional[str] = None
    cl_jg: Optional[str] = None
    wx_yy: Optional[str] = None
    wx_nr: Optional[str] = None

@risk_fault_api.post("/update_risk_fault")
async def update_risk_fault(info_data: riskfaultUpdateModel):
    try:
        info = await risk_fault_records.get_or_none(id=info_data.id)
        if not info:
            return {"status": 1, "msg": f"未找到ID为{info_data.id}的记录"}
        
        await info.update_from_dict(info_data.dict(exclude_unset=True))
        await info.save()
        return {"status": 0, "msg": "更新成功"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"更新失败: {str(e)}")

@risk_fault_api.delete("/delete_risk_fault/{id}")
async def delete_risk_fault(id: int):
    try:
        info = await risk_fault_records.get_or_none(id=id)
        if not info:
            return {"status": 1, "msg": f"未找到ID为{id}的记录"}
        
        await info.delete()
        return {"status": 0, "msg": "删除成功"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"删除失败: {str(e)}")